Ical.Net/CalendarExtensions.cs

```diff
@@ -11,20 +11,32 @@ namespace Ical.Net;
 public static class CalendarExtensions
 {
     /// <summary>
-    /// https://blogs.msdn.microsoft.com/shawnste/2006/01/24/iso-8601-week-of-year-format-in-microsoft-net/
+    /// Calculate the week number according to ISO.8601, as required by RFC 5545.
     /// </summary>
-    public static int GetIso8601WeekOfYear(this System.Globalization.Calendar calendar, DateTime time, CalendarWeekRule rule, DayOfWeek firstDayOfWeek)
+    public static int GetIso8601WeekOfYear(this System.Globalization.Calendar calendar, DateTime time, DayOfWeek firstDayOfWeek)
     {
-        // Seriously cheat.  If its Monday, Tuesday or Wednesday, then it'll
-        // be the same week# as whatever Thursday, Friday or Saturday are,
-        // and we always get those right
-        var day = calendar.GetDayOfWeek(time);
-        if (day >= DayOfWeek.Monday && day <= DayOfWeek.Wednesday)
-        {
-            time = time.AddDays(3);
-        }
+        // A week is defined as a
+        // seven day period, starting on the day of the week defined to be
+        // the week start(see WKST). Week number one of the calendar year
+        // is the first week that contains at least four (4) days in that
+        // calendar year.
 
-        // Return the week of our adjusted day
-        return calendar.GetWeekOfYear(time, rule, firstDayOfWeek);
+        // We add 3 to make sure the test date is in the 'right' year, because
+        // otherwise we might end up with week 53 in a year that only has 52.
+        var tTest = GetStartOfWeek(time, firstDayOfWeek).AddDays(3);
+        var res = calendar.GetWeekOfYear(tTest, CalendarWeekRule.FirstFourDayWeek, firstDayOfWeek);
+
+        return res;
+    }
+
+    /// <summary>
+    /// Calculate and return the date that represents the first day of the week the given date is
+    /// in, according to the week numbering required by RFC 5545.
+    /// </summary>
+    private static DateTime GetStartOfWeek(this DateTime t, DayOfWeek firstDayOfWeek)
+    {
+        var t0 = ((int) firstDayOfWeek) % 7;
+        var tn = ((int) t.DayOfWeek) % 7;
+        return t.AddDays(-((tn + 7 - t0) % 7));
```

Comment:
Very well explained

Reply:
Thank you. Actually its a citation from RFC 5545 ;-)